### DISCUSSION TITLE
[FR] frontend ui development planning

### DISCUSSION THREAD

**Discussion by ljmerza** [blakeblackshear/frigate#638] [Ideas]
Now that the frontend is being built out I think it's a good idea to step back and make some design choices before development on it continues. The main problem i see here is there's no centralized color scheme so say we want to add a dark theme it would be very hard to do so without a bunch of rewriting. I work with React a lot so i think it's a great choice for the ui. i think it would be helpful to pick out a ui library that supports colors schemes. maybe we can get a separate project roadmap going for the ui to plan out long term features and design the ui with those in mind?

**Comment by sakalauskas** [comment]
I think Tailwind (And React-style framework) is a great choice here (I work with them a lot as well).

Actually, the dark mode can be quite easily added:
https://tailwindcss.com/docs/dark-mode

**Comment by paularmstrong** [comment]
> so say we want to add a dark theme

There is a dark theme. Tailwind chooses it based on the user system preference.
<img width="1348" alt="Screen Shot 2021-01-27 at 7 32 34 PM" src="https://user-images.githubusercontent.com/33297/106086139-6c3ae900-60d6-11eb-8114-f226a9ffb177.png">
<img width="1392" alt="Screen Shot 2021-01-27 at 7 32 25 PM" src="https://user-images.githubusercontent.com/33297/106086147-6e9d4300-60d6-11eb-884e-4d8cec3b0732.png">

The "color scheme" is using Tailwind's default theme of grays with blue for links.

**Comment by ljmerza** [comment]
The color scheme was more of an example. Overall I'd like to see frontend roadmaps to plan out development.

**Reply by paularmstrong** [reply]
In your original post, you said that's the "main problem", so I was responding to that.

> main problem i see here is there's no centralized color scheme

If that's not all there is, what else did you have in mind? Everything so far has been pretty ad-hoc, just to get something working instead of some plain text with no navigation. Might help to make a list of the things you'd like to see added or changed and maybe open some pull requests. I'm happy to help out where I can.

**Comment by ljmerza** [comment]
I assumed there were already ideas of what the overall goal of the UI was. We could gather a preliminary list in this discussion. I'll start with a few:

- customizing settings via the ui is a standard but probably low on the priority list right now. 
- turning on/off cameras/recording/clips per camera
- loading spinner
- event filtering/sorting. i see there's a round about way to do it but it would be nice to have (multi?) dropdowns selecting cameras, labels,
- alerts through things like pushbullet, discord ,etc. right now i use a hacky nodered flow for notifications. also be nice to turn on/off notifications at certain times.
-  not everyone uses home assistant so some kind of file management?
- users and logging in. this would be a big one. maybe admin user vs regular user? how to handle security.


I'll try to add some more later on.

**Reply by paularmstrong** [reply]
Most of this will require API changes, as Blake notes below. 

@blakeblackshear would you mind if I created a separate "Project" in this repo and created & tagged issues onto that roadmap? We can try to prioritize them via 👍🏻  votes and perhaps you could close any that you don't think fits into your vision for Frigate.

**Reply by blakeblackshear** [reply]
Go for it.

**Reply by paularmstrong** [reply]
> * turning on/off cameras/recording/clips per camera

#647

> * loading spinner

* #601
* #648
* #649

> * event filtering/sorting. i see there's a round about way to do it but it would be nice to have (multi?) dropdowns selecting cameras, labels,

* #601 
* #650

**Reply by paularmstrong** [reply]
Project is here: https://github.com/blakeblackshear/frigate/projects/2
I've assigned a few to myself that I'd like to make headway on in the next day or two.

**Reply by blakeblackshear** [reply]
Everything there looks great to me.

**Comment by blakeblackshear** [comment]
> customizing settings via the ui is a standard but probably low on the priority list right now.

I am most interested in having a "configuration builder" as something like a setup wizard for adding a camera. This will always need to function as a method to generate a yaml configuration. In my personal setup, frigate will never be able to write to the config file since I manage it in source control and deploy with kubernetes as a read only file in the container. For those with the ability to write to the config file, we could add the ability to save over it and force a restart of the container.

> turning on/off cameras/recording/clips per camera

To the extent that you can turn things on or off via mqtt, we can add api endpoints and enable it in the UI too.

> loading spinner

Sure, but where exactly? The UI is instant for me.

> event filtering/sorting. i see there's a round about way to do it but it would be nice to have (multi?) dropdowns selecting cameras, labels,

The API supports filtering by a variety of things already, so this seem straightforward.

> alerts through things like pushbullet, discord ,etc. right now i use a hacky nodered flow for notifications. also be nice to turn on/off notifications at certain times.

This isn't a UI feature. I have no interest in reinventing the notifications wheel. I would rather make the changes necessary to create a "simple" nodered flow or homeassistant automation. Those platforms already support integrations with hundreds of notification services.

> not everyone uses home assistant so some kind of file management?

Homeassistant isnt doing any file management. I'm not sure what you mean by file management.

> users and logging in. this would be a big one. maybe admin user vs regular user? how to handle security.

This will never be a part of frigate. If you want authentication, run frigate behind Traefik with traefik-forward-auth, Nginx with oauth2_proxy, or use Homeassistant.

**Reply by paularmstrong** [reply]
> I am most interested in having a "configuration builder" as something like a setup wizard for adding a camera. This will always need to function as a method to generate a yaml configuration. In my personal setup, frigate will never be able to write to the config file since I manage it in source control and deploy with kubernetes as a read only file in the container. For those with the ability to write to the config file, we could add the ability to save over it and force a restart of the container.

Going out on a limb, I'm going to guess that you're in the minority of wanting this. I absolutely respect that this is _your_ project and ultimately, the direction is chosen by you, but I would like to put a +1 on being able to do all the configuration from a UI. YAML is a beast, especially with the amount of configuration both required and optional that's available for Frigate. Makes it pretty hard for people getting set up and ends up requiring a lot of "tweak, restart, check, tweak, restart..." to get things right. Running Frigate as a HA addon, my configuration is backed up with the rest of the HA daily/weekly backup, so I don't really worry about keeping it somewhere else.

**Reply by blakeblackshear** [reply]
Absolutely agree, but I see no reason why both things cannot be accomplished at the same time. I am imagining the frontend as essentially a tool to build/modify the json (yaml is just json) for the config with an endpoint that accepts the config and updates it on disk. I just want to make sure that the generated config can also be copy/pasted from the frontend. I am all for adding endpoints to frigate that assist in adding a new camera such as adding an input path and grabbing a still frame to make sure it works, modifying input params, and even some kind of a device registry with know working ffmpeg params.

**Reply by ljmerza** [reply]
i would love to replace blue iris with frigate. i see no point in having two apps that record and track motion but frigate doesnt have the features yet to switch over. not saying frigate should be as big as blue iris but considering how blue iris is still stuck on windows, i can see frigate really blowing up if a good amount of features are added into the ui

**Comment by Ysbrand** [comment]
> > loading spinner
> Sure, but where exactly? The UI is instant for me.

In the HA add-on it can easily take 15 seconds after selecting "events" before the results are displayed.

**Reply by paularmstrong** [reply]
This can happen if you have 100+ events, because the current query limit is 100. I have an in-progress change to make this only load 10 at a time and paginate, but it needs more thought, because it's not possible with the current API to link to a particular page of events, nor go past the 100 event limit

**Reply by blakeblackshear** [reply]
There is a limit param on that endpoint too, but not paging. You may also want to pass `include_thumbnails=false` to reduce the payload size and use the `thumbnail.jpg` endpoints fo the images so they can be loaded on demand.

**Reply by ljmerza** [reply]
we should have a loading spinner for all api calls. its a simple adjustment that makes an app look far more polished. we cant assume how fast or slow anyone's connection is nor the device they are using. for example, if someone uses nginx to allow for remote access over mobile then everything will need a loading spinner. or someone uses the pi3, there will be some slow downs. maybe there's something else running at the same time like an event? too many variables. always assume loading time.

**Reply by paularmstrong** [reply]
Yes, there are a couple places where a loading state should be given if it's taking too long. I'd like to avoid that as much as possible, as the app is already incredibly fast. The main issue at the moment is timing of API calls and lack of persistent state outside of the main configuration.

**Reply by paularmstrong** [reply]
> There is a limit param on that endpoint too, but not paging. You may also want to pass `include_thumbnails=false` to reduce the payload size and use the thumbnail.jpg endpoints fo the images so they can be loaded on demand.

Looks like this parameter got removed at some point. It will have to get added back in if we need it.

**Comment by paularmstrong** [comment]
@ljmerza Thanks for all the ideas on the UI. A bunch of stuff is going to make it out into the [v0.8.1 release](https://github.com/blakeblackshear/frigate/milestone/1?closed=1) and I'm working on a couple for [v0.8.2](https://github.com/blakeblackshear/frigate/milestone/2). 

You and anyone else should feel free to pick up any of the issues that I haven't already assigned to myself!
